src/include/function/binary_function_executor.h

```diff
@@ -110,6 +110,7 @@ struct BinaryFunctionExecutor {
         if (left.isNull(lPos)) {
             result.setAllNull();
         } else if (right.hasNoNullsGuarantee()) {
+            result.setAllNonNull();
```

Comment:
Should the null mask be updated here or should be reset it higher up (e.g. in `FunctionExpressionEvaluator::evaluate()` or in the operator's `getNextTuplesInternal()` function)?

Reply:
Was this causing bug? The `executeOnValue` never read from `result nullMask` but instead always read from `left & right nullMask` and write to `result nullMask`. So the previous implementation should be safe.

Reply:
The problem was that this branch doesn't modify the result nullmask so something like below can happen:
1. Perform vector comparison, result at pos[0] is null, result nullmask correctly updated so that pos[0] is null
2. Perform vector comparison, result should have no nulls (children are both non-null). We reuse the result nullmask from step 1 but never update it so pos[0] is still reported as null

Reply:
sounds good. Then I would expect test cases to reproduce this bug.

Reply:
ah right thanks for the reminder I'll add them to this PR

Reply:
added test